redis_collections/lists.py

```diff
@@ -230,26 +275,57 @@ def __delitem__(self, index):
         begin = 0
         end = -1
 
+        # Calculate the cache index before the length changes
+        if self.writeback:
+            cache_index = self._get_cache_index(index)
+
         if isinstance(index, slice):
-            self._del_slice(index)
+            return self._del_slice(index)
+
+        if index == begin:
+            self.redis.lpop(self.key)
+        elif index == end:
+            self.redis.rpop(self.key)
         else:
-            if index == begin:
-                self.redis.lpop(self.key)
-            elif index == end:
-                self.redis.rpop(self.key)
-            else:
-                raise NotImplementedError(self.not_impl_msg)
+            raise NotImplementedError(self.not_impl_msg)
+
+        # Removing an item from the list means all the other items after it
+        # have to shift back one - reflect that in the cache
+        if self.writeback:
+            new_cache = {}
+            for k, v in six.iteritems(self.cache):
+                if k < cache_index:
+                    new_cache[k] = v
+                elif k == cache_index:
+                    pass
+                elif k > cache_index:
+                    new_cache[k - 1] = v
+
+            self.cache = new_cache
 
     def remove(self, value):
         """Remove the first occurence of *value*."""
-        self.redis.lrem(self.key, 1, self._pickle(value))
+        # If we're caching, we'll need to synchronize before removing.
+        with self.redis.pipeline() as pipe:
+            if self.writeback:
+                self._sync_helper(pipe)
```

Comment:
This is a punt; I started trying a "set to marker, then remove the marker" procedure, but it got cumbersome. I will probably come back and change this later if I implement more Python slicing.